Add C++ as a language option

WhyPenguins opened the pull request:
# Description

This PR adds support for compiling and running C++ inside SplashKit Online. It is not feature complete - audio and filesystem access does not work. For the most part however it works quite well, and so I think it's still worth merging in now, with follow-up PRs coming for finishing up full feature support.

## Testing

### Getting the binaries
In order to test this, you'll need binaries for the following:
 - C++ Compiler
 - C++ Compiler System Root Files
 - C++ Runtime
 - JS Runtime (for usual JavaScript support - not needed for testing C++)

These can all be found in the branch [here](https://github.com/WhyPenguins/SplashkitOnline/tree/cxx_language_backend_binaries) - just copy the files into the same spot in your local repo.

- The C++ compiler and pre-built system root files can be found in `Browser_IDE/compilers/cxx/bin/*`
  - You'll need to extract `compiler.zip` - I just had to compress the files in it to stay within GitHub's file size limit.
- The pre-built C++ runtime can be found in `Browser_IDE/runtimes/cxx/bin/*`
- The pre-built JavaScript runtime can be found in `Browser_IDE/runtimes/javascript/bin/*`

### Compiling the binaries
If you want to compile some of this yourself, currently the repository supports building both the C++/JavaScript runtimes, and the Compiler System Root Files (partially). Building the compilers themselves is not yet supported - I intend to make this reproducible soon though. To compile the C++ system root file, you'll need the following file:  `SplashKitWasm/prebuilt/sysroot.zip`
From there you can follow the instructions on the README, just use the following command instead:
```bash
cd SplashkitOnline\SplashKitWasm\cmake
emcmake cmake -G "Unix Makefiles" -DENABLE_CPP_BACKEND=ON .
emmake make -j8
```
(Compiling the JavaScript backend is enabled by default, while the C++ one currently needs to be enabled with `ENABLE_CPP_BACKEND=ON`. I also recommend the `-j8` to speed up compilation, might cause trouble depending on machine specs though)

### Switching to C++

Once its all setup, you can head to `localhost:8000`, which will bring up the normal JavaScript runtime. To switch to C++, just go to `localhost:8000/?language=CXX` - you'll probably also want to click `New Project` to create the default C++ project. You can also use `localhost:8000/?language=JavaScript` for JavaScript.

## Brief explanation of how this works

There's now a new file - `languageDefinitions.js` that defines the languages SplashKit supports. The languages are basically just a list of script files to include, either for compilation or the runtime.

### Compilation
C++ compilation and linking is all just done using Clang compiled to WebAssembly. This is actually pretty straightforward - the user's source files are written into `clang++`'s filesystem, and clang is invoked to compile them. These object files are read out, and then written into `wasm-ld` (the linker)'s filesystem, where it links them into a `.wasm` (WebAssembly), which we can then instantiate and run. High level work is handled in `Browser_IDE/compilers/cxx/cxxCompiler.js`, and more detailed work in `Browser_IDE/compilers/cxx/cxxCompilerClangBackend.js`

Couple of tricky things:
 - Instantiating the compiler modules takes up a lot of RAM (~1gb). Doing this repeatedly ends up leaking RAM (on the browsers I tested anyway) causing the process to fail, so we want to re-use a single instantiation instead.
 - Repeatedly calling the compiler's `main` _works_, for about 8 tries. Clang internally has a signal array that fills up after 8 runs, afterwhich it starts aborting.
 - So we just tell it to clean it up by pretending we've SIGKILL'd it, then continue calling its main.
 - We also run the compiler inside a WebWorker (another thread basically, `Browser_IDE/compilers/cxx/cxxCompilerClangWebWorker.js`) - this way compilation doesn't block the main interface. It also has a positive that I've found WebWorkers clean up their memory better than the main page, so repeated refreshes no longer make the browser crash 😄 

### Runtime
The runtime (`Browser_IDE/runtimes/cxx/cxxRuntime.js`) just (ab)uses the Emscripten generated JavaScript to run the user's compiled `.wasm` rather than the one Emscripten generated the glue code for. This is a bit of a hack, but it works surprisingly well. To make it work we have to patch Emscripten's generated glue a bit - this is just done with a Python script as a post process (`SplashKitWasm/tools/cxx_backend_emscripten_glue_patcher.py`).

Tricky things:
 - If the user's code contains a while loop, we'll block the page's main thread and the page will freeze.
 - Solution: Run the user's code in a WebWorker! WebWorker's can't draw on the canvas, test for input or play sound, but we can ask Emscripten to generate proxying code for us - it'll send messages to the WebWorker on inputs, and the WebWorker sends drawing commands and such back.
 - Problem: If the user's code contains a while loop, the WebWorker can't receive messages! It can _send_ them, but couldn't receive input events.
 - Solution A) Use `SharedArrayBuffer`s, which allow the main page and WebWorker to share an array - then just read from that array in SplashKit's `process_events` to check for events. This works well, but it requires special HTTP headers we can't set on GitHub (so we couldn't host it on GitHub anymore), and has very iffy browser support.
 - Solution B) WebWorkers can make web requests - so we set up a 'Service Worker' (`Browser_IDE/SKOservice-worker.js`) in between the main page and Web Worker, which acts as a proxy. We send events to it for queuing, and then when the WebWorker requests the file '/programEvents.js', we give it a JSON file filled with the latest events! (see `Browser_IDE/runtimes/cxx/workerEventProcessor.js`) This is what is used in this PR.

And that's the gist of it!

## List of changes
 - New WASM library compilation targets (for JS and C++)
 - Language definitions that control which scripts are loaded
 - C++ compiler backend
 - JavaScript 'compiler' wrapper (needs refactoring later)
 - C++ runtime based on asynchronous execution via WebWorker (inside Execution Environment iFrame)
 - Service worker to handle communication between C++ runtime and Execution Environment
 - Project initializer for C++
 - Small improvements to notifications
 - Fallible messages extension (can now resolve/reject like promises)
 - Terminal styling improvement (monospace + retain whitespace)
 - Error handling for programRun/Pause etc
 - Added favicon

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

C++ compilation and execution with the new default project has been tested on both desktops and mobile, on Firefox, Edge and Opera. Error reporting has been tested by adding random syntax errors - runtime errors have not been tested. 
JavaScript execution has also been checked for regressions, with the default project and Cave Escape.

## Testing Checklist

- [x] Tested in latest Edge (basically Chrome)
- [ ] Tested in latest Safari
- [x] Tested in latest Firefox

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have requested a review from ... on the Pull Request

WhyPenguins commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add C++ language option](https://trello.com/c/CC6edvT8/576-add-c-language-option)

jessbdeakin commented on Browser_IDE/splashKitOnlineEnvParams.js:
```diff
@@ -0,0 +1,18 @@
+"use strict";
+
+function findGetParameter(parameterName, _default=null) {
+    location.search.substr(1).split("&").forEach(function (piece) {
+      let tmp = piece.split("=");
+      if (tmp[0] === parameterName)
+          return decodeURIComponent(tmp[1]);
+    });
+
+    return _default;
+}
```
I can't find any place in which this function is actually used. Am I missing one?

WhyPenguins replied:
It _was_ used... You can probably tell that `splashKitOnlineEnvParams.js` was a bit rushed 😅

jessbdeakin commented on Browser_IDE/splashKitOnlineEnvParams.js:
```diff
@@ -0,0 +1,18 @@
+
+// global object that can be used to configure the IDE
+let SKO = {
+	language: page_url.searchParams.get("language") ?? "JavaScript",
```
URLSearchParams (searchParams) converts plus characters to spaces when interpreting the URL query string, so 'C++' will not actually work as a parameter value, and the error is somewhat abstruse: 'Unable to switch to language C , defaulting to JavaScript.'

jessbdeakin reviewed (commented):
This is obviously still a WIP feature so I have tried not to note things you are likely already aware of. The compiler and runtime code seem rather clean, and I have not noticed anything erroneous on a line-by-line basis. Nothing has regressed, with one exception.

I encountered a bug while testing that is somewhat difficult to reproduce. If while the user's program is running, the execution of the page is paused with the browser's debugger for long enough, the page will lock up entirely upon unpausing. My guess is that this is somehow related to the workers, as while the page is frozen, I am still able to see requests for `programEvents.js` being sent. Perhaps when the page unpauses, it is suddenly inundated with thousands of yet unhandled requests? I cannot think of any conditions under which this would trigger for a normal user, but it would make development more difficult.

This is less immediately relevant but: I find the notifications slightly cluttered. Immediate feedback is an important part of a tool like SplashKit Online, but now if the user restarts the program more than 3~4 times in short succession, the entire terminal and canvas are obscured by notifications. Some of these messages might be better suited to a status bar; one is already carved out at the bottom of the page, though goes unused.

WhyPenguins commented:
Thanks so much for reviewing this! Really glad to know someone else was able to get it setup and working - did the instructions cover everything or were there issues? If they were alright I'll add a tidied up version of these instructions to the README too as part of the PR.

> I encountered a bug while testing that is somewhat difficult to reproduce. If while the user's program is running, the execution of the page is paused with the browser's debugger for long enough, the page will lock up entirely upon unpausing. My guess is that this is somehow related to the workers, as while the page is frozen, I am still able to see requests for `programEvents.js` being sent. Perhaps when the page unpauses, it is suddenly inundated with thousands of yet unhandled requests? I cannot think of any conditions under which this would trigger for a normal user, but it would make development more difficult.

Ah that's a good one, I'll have a look and see if I can reproduce it.

> This is less immediately relevant but: I find the notifications slightly cluttered. Immediate feedback is an important part of a tool like SplashKit Online, but now if the user restarts the program more than 3~4 times in short succession, the entire terminal and canvas are obscured by notifications. Some of these messages might be better suited to a status bar; one is already carved out at the bottom of the page, though goes unused.

Yes I was wondering about this too. As a stop-gap for this PR I might just make the notifications vanish after 1.5 seconds rather than 5, since this will largely mitigate the problem - does this seem alright?

As another task, I was thinking of either adding a status bar like you suggest, or perhaps making the notifications themselves update (so a single notification might appear for 'Compiling...', perhaps with a progress bar, and then once compilation completes it would update to say 'Running project!', and then vanish after the usual few seconds).

jessbdeakin commented:
I didn't face any issues during set-up, no. The instructions are perfectly clear.

WhyPenguins commented on Browser_IDE/languageDefinitions.js:
```diff
@@ -0,0 +1,72 @@
+    }
+};
+
+// Some synonyms to make development easier
+SplashKitOnlineLanguageDefinitions["CXX"] = SplashKitOnlineLanguageDefinitions["C++"];
+SplashKitOnlineLanguageDefinitions["C"] = SplashKitOnlineLanguageDefinitions["C++"];
+SplashKitOnlineLanguageDefinitions["JS"] = SplashKitOnlineLanguageDefinitions["JavaScript"];
```
Ah that's a good point - the aliases probably belong inside the language definitions themselves anyway (this was a bit of a hack really...)

WhyPenguins commented:
@jessbdeakin Alright I think I've addressed all the issues - fixed the debugging problem too 😄 Just going to update the README as well and then if it's okay with you I'll go ahead and merge it!

WhyPenguins commented:
Alright README is updated too, so should be good to go hopefully.

jessbdeakin commented:
Though this is already merged, I nevertheless wanted to note that these changes all look great. I am excited to build on this feature.
